Staging/palma dev

danmois opened the pull request:
## PR Description

The ADRV9032R is a highly integrated, RF agile transceiver offering two transmitters, two observation receivers for monitoring transmitter channels, two receivers, integrated local oscillator (LO) and clock synthesizers, and digital-signal processing functions to provide a complete transceiver solution. The device provides the high radio performance and low-power consumption demanded by cellular infrastructure applications, software-defined radios, portable instruments, and military communications.

The receiver and transmitter signal paths use a zero-IF (ZIF) architecture that provides wide bandwidth with dynamic range suitable for non-contiguous multicarrier applications. The ZIF architecture has the benefits of low power and RF and bandwidth agility. The lack of aliases and out-of-band images eliminates anti-aliasing and image filters, reducing system size and cost, and making band independent solutions possible.

## PR Type
- [ ] Bug fix (a change that fixes an issue)
- [x] New feature (a change that adds new functionality)
- [ ] Breaking change (a change that affects other repos or cause CIs to fail)

## PR Checklist
- [x] I have conducted a self-review of my own code changes
- [x] I have compiled my changes, including the documentation
- [x] I have tested the changes on the relevant hardware
- [ ] I have updated the documentation outside this repo accordingly
- [ ] I have provided links for the relevant upstream lore

danmois committed:
iio: trx-rf: adrv903x: Add ADRV903X API

Add the API Version 2.12.1.4 for ADRV903X.

Signed-off-by: George Mois <george.mois@analog.com>

danmois committed:
firmware: Add ADRV9030 firmware

Add ADRV9030 binaries and gain tables.

Signed-off-by: George Mois <george.mois@analog.com>

danmois committed:
drivers: iio: Kconfig.adi: Imply ADRV903X

Make sure that the ADRV903X driver is built.

Signed-off-by: George Mois <george.mois@analog.com>

danmois committed:
arch: arm64: dts: Add ADRV9032R dts.

Add the dts file for the ADRV9032R development board:
arch/arm64/boot/dts/xilinx/zynqmp-zcu102-rev10-adrv9032r-nls.dts.

Signed-off-by: George Mois <george.mois@analog.com>

danmois committed:
iio: adc: ad_adc: Add ORX info for ADRV903X

Add info table for observation receiver for the ADRV903X family.

Signed-off-by: George Mois <george.mois@analog.com>

stefpopa committed:
iio: trx-rf: adrv903x: fix VLA in vendor API HealthMonitorCpuStatusGet

Fix variable length array compilation error in ADI vendor API code
that triggers build failure with -Werror=vla. The VLA was caused by
using sizeof() in array dimension, which the compiler cannot evaluate
as a constant expression.

Replace with a fixed 32-word buffer (128 bytes) to accommodate the
66-byte HealthMonitorCpuStatus structure with headroom for future
changes. Add runtime validation to ensure buffer size is sufficient.

Signed-off-by: Stefan Popa <stefan.popa@analog.com>

stefpopa committed:
dt-bindings: iio: adc: Add ADRV903x RF transceiver bindings

Add device tree bindings documentation for the Analog Devices
ADRV9032/ADRV9032R RF transceiver devices.

File:
- Documentation/devicetree/bindings/iio/adc/adi,adrv903x.yaml

Signed-off-by: Stefan Popa <stefan.popa@analog.com>

gastmaier commented on drivers/iio/adc/adrv903x/adrv903x.c:
```diff
@@ -0,0 +1,3409 @@
+	case TX_SAMPL_CLK:
+		adrv903x_TxLinkSamplingRateFind(phy->palauDevice,
+						ADI_ADRV903X_DEFRAMER_0,
+						&rate);
```
rate may be uninitialized if at adrv903x_TxLinkSamplingRateFind device is null or

```
		ADI_ERROR_REPORT(&device->common, ADI_COMMON_ERRSRC_API,
				 ADI_COMMON_ERRCODE_INVALID_PARAM,
				 ADI_COMMON_ERR_ACT_CHECK_PARAM, deframerSel,
				 "Only one deframer can be selected at a time.");
		if (device->common.errPtr->errDebugInfo.highestPriorityAction)
			return device->common.errPtr->errDebugInfo.highestPriorityAction;
```

gastmaier commented on arch/arm64/boot/dts/xilinx/zynqmp-zcu102-rev10-adrv9032r-nls.dts:
```diff
@@ -0,0 +1,438 @@
+/*
+ * Analog Devices ADRV903X
+ * https://wiki.analog.com/resources/eval/user-guides/adrv903x
+ * https://wiki.analog.com/resources/tools-software/linux-drivers/iio-transceiver/adrv903x
```
Those links does not exist
what is the documentation plan for adrv903x?

stefpopa replied:
I added a link that exists, but will create github.io documentation page for this part: https://analogdevicesinc.github.io/documentation/ ... WIP

gastmaier reviewed (changes requested):
Quick questions + improvement request

stefpopa commented on Documentation/devicetree/bindings/iio/adc/adi,adrv903x.yaml:
```diff
@@ -0,0 +1,161 @@
+  reg:
+    maxItems: 1
+
+  spi-max-frequency: true
```
fixed

nunojsa commented on drivers/iio/adc/adrv903x/platforms/adi_library_types.h:
```diff
@@ -54,27 +54,55 @@ typedef time64_t time_t;
+#else
+#define ADI_LIBRARY_CALLOC                              __adrv903x_calloc
+#define ADI_LIBRARY_FREE                                kfree
+#endif
```
I would be nice if all these preparation changes (including the linux platform) were done in a separate commit from the main drivers. Same for cf_axi_dds change

nunojsa commented on drivers/iio/adc/adrv903x/adrv903x.c:
```diff
@@ -0,0 +1,3412 @@
+ *
+ * Copyright 2020-2025 Analog Devices Inc.
+ *
+ * Licensed under the GPL-2.
```
drop the above. No need for it

nunojsa commented on drivers/iio/adc/adrv903x/adrv903x.c:
```diff
@@ -0,0 +1,3412 @@
+
+static const struct spi_device_id adrv903x_id[] = {
+	{ "adrv9032", ID_ADRV9032 },
+	{ "adrv9032r", ID_ADRV9032R },
```
We have ID here and not in the `of_device_id`. Not consistent. But more on that in a moment

nunojsa commented on drivers/iio/adc/adrv903x/adrv903x.c:
```diff
@@ -0,0 +1,3412 @@
+out_clk_del_provider:
+	of_clk_del_provider(np);
+out_disable_clocks:
+	clk_disable_unprepare(phy->dev_clk);
```
use devm_ variants. No need for the above. I mean, there's no .remove() as well

nunojsa reviewed (commented):
Here it goes my initial review. We need to take some care in modernizing these drivers (which are fairly complex). Like this, it makes a bit harder to review.

stefpopa committed:
iio: frequency: cf_axi_dds: Fix ERR_PTR dereferences and off-by-one

Fix missing IS_ERR() checks on the return value of to_converter()
in cf_axi_get_parent_sampling_frequency(), cf_axi_dds_sync_frame() and
cf_axi_dds_read_raw(). The pointer was dereferenced unconditionally,
which is wrong since to_converter() can return an ERR_PTR.

Also fix an off-by-one in cf_axi_dds_setup_chip_info_tbl(): the bounds
check used '>' instead of '>=' against ARRAY_SIZE(), allowing one
element past the end of the channel array to be written.

Signed-off-by: Stefan Popa <stefan.popa@analog.com>

stefpopa committed:
iio: frequency: cf_axi_dds: Add ADRV903X support

Add adrv903x_1_00_a_info core info and the "adi,axi-adrv903x-tx-1.0"
compatible string to support the ADRV903X DAC.

Signed-off-by: Stefan Popa <stefan.popa@analog.com>

stefpopa committed:
iio: trx-rf: adrv903x: Add Linux platform adaptation layer

Add the Linux kernel platform layer that adapts the ADI vendor library
to the kernel environment.

adi_library_types.h is updated with __KERNEL__ guards to redirect
standard C library calls (printf, malloc, fopen, etc.) to their kernel
equivalents (printk, kmalloc, kernel file I/O wrappers).

linux_platform.c implements the HAL functions required by the ADI
library: SPI read/write via the kernel SPI subsystem, memory allocation
via kmalloc/kfree, and file I/O stubs for firmware loading.

Signed-off-by: Stefan Popa <stefan.popa@analog.com>

stefpopa committed:
iio: trx-rf: Add ADRV903X support

Add driver files for ADRV903X and add option to makefiles.

Signed-off-by: George Mois <george.mois@analog.com>
Signed-off-by: Stefan Popa <stefan.popa@analog.com>

nunojsa commented on Documentation/devicetree/bindings/iio/adc/adi,adrv903x.yaml:
```diff
@@ -0,0 +1,163 @@
+  adi,rx-gaintable-names:
+    description:
+      Filename for receiver gain table (CSV format)
+    $ref: /schemas/types.yaml#/definitions/string
```
Do we have a fixed name for those? Or is it flexible?

nunojsa commented on drivers/iio/adc/adrv903x/adrv903x.c:
```diff
@@ -0,0 +1,3406 @@
+						 &phy->trxBinaryInfoPtr);
+	if (recoveryAction != ADI_ADRV903X_ERR_ACT_NONE) {
+		pr_err("ERROR adi_adrv903x_PreMcsInit failed in %s at line %d.\n", __func__,
+		       __LINE__);
```
These plain pr_err() are odd

nunojsa commented on drivers/iio/adc/adrv903x/adrv903x.c:
```diff
@@ -0,0 +1,3406 @@
+
+	switch (entry->cmd) {
+	case DBGFS_BIST_FRAMER_0_PRBS:
+		mutex_lock(&phy->lock);
```
Maybe just lock the whole function with guard()

nunojsa reviewed (commented):
Ok, I think it's already a lot for now :)

nunojsa reviewed (commented):
Just some minor stuff. Given that main was updated with apollo, it would make sense to also put this device under:

https://github.com/analogdevicesinc/linux/tree/main/drivers/iio/trx-rf

Given this is not an ADC :)

stefpopa commented:
@nunojsa thank you for your review! Here's what changed:

- Moved driver to drivers/iio/trx-rf
- Git subject: no .c in ad_adc.c commit
- Makefile: use $(src) for out-of-tree builds
- adrv903x_register_debugfs must not fail probing
- Copyright year in adrv903x_conv.c
- Add Stefan Popa as MODULE_AUTHOR
- { } sentinel entries in of_match/spi_id
- No forced line-breaks for _parent_name assignments

nunojsa commented on drivers/iio/trx-rf/adrv903x/common/adi_error/adi_common_error.c:
```diff
@@ -0,0 +1,324 @@
+ */
+
+#include "adi_common_error.h"
+#include "adi_platform_types.h"
```
Am I hallucinating or is this still on iio/adc?

stefpopa replied:
You are right. I have just moved it.

nunojsa commented on drivers/iio/adc/Makefile:
```diff
@@ -117,6 +117,7 @@ obj-$(CONFIG_ADRV9009) += adrv9009_drv.o
 
 obj-$(CONFIG_ADRV9025) += adrv902x/
 
+
```
Ok, this commit is moving but it does not make sense to do it like this 😅. It should be in the right path from the beginning. I don't remember exactly the steps I've done during rebase but ask claude. It helped on the same problem with apollo

stefpopa replied:
Good point. Done.

stefpopa commented:
changelog:
- Path fix: move adrv903x to iio/trx-rf from the start
- Stray blank lines removed

nunojsa commented on arch/arm64/configs/adi_zynqmp_defconfig:
```diff
@@ -113,7 +113,7 @@ CONFIG_PCI=y
 CONFIG_DEVTMPFS=y
 CONFIG_DEVTMPFS_MOUNT=y
-CONFIG_EXTRA_FIRMWARE="ad9467_intbypass_ad9517.stp ad9517.stp ad9144_fmc_ebz_ad9516.stp Mykonos_M3.bin TaliseStream.bin TaliseTDDArmFirmware.bin TaliseTxArmFirmware.bin TaliseRxArmFirmware.bin adau1761.bin Navassa_EvaluationFw.bin RxGainTable.csv RxGainTable_GainCompensated.csv ORxGainTable.csv TxAttenTable.csv Navassa_Stream.bin Navassa_CMOS_profile.json Navassa_LVDS_profile.json Navassa_CMOS_pr ...
+CONFIG_EXTRA_FIRMWARE="ad9467_intbypass_ad9517.stp ad9517.stp ad9144_fmc_ebz_ad9516.stp Mykonos_M3.bin TaliseStream.bin TaliseTDDArmFirmware.bin TaliseTxArmFirmware.bin TaliseRxArmFirmware.bin adau1761.bin Navassa_EvaluationFw.bin RxGainTable.csv RxGainTable_GainCompensated.csv ORxGainTable.csv TxAttenTable.csv Navassa_Stream.bin Navassa_CMOS_profile.json Navassa_LVDS_profile.json Navassa_CMOS_pr ...
```
Can you move this to a different patch? Very handy on rebased branches

stefpopa replied:
done

nunojsa commented on drivers/iio/trx-rf/adrv903x/Makefile:
```diff
@@ -69,5 +69,13 @@ ccflags-y += -I$(src)/devices/adrv903x/private/include/ \
+CFLAGS_devices/adrv903x/public/src/adi_adrv903x_utilities.o := -Wframe-larger-than=2048
+
+# Suppress format warnings for vendor error handling (uses C standard PRIx32 macros incompatible with printk)
+CFLAGS_common/adi_error/adi_common_error.o := -Wno-format
```
Arguably, I would squash this in the patch adding the API

stefpopa replied:
done

nunojsa reviewed (commented):
Looks like there are valid checkpatch complains:

https://github.com/analogdevicesinc/linux/actions/runs/24028797400/job/70072994872?pr=3060

stefpopa committed:
arch: arm64: configs: Add ADRV9030 firmware to adi_zynqmp_defconfig

Add ADRV9030 firmware files to CONFIG_EXTRA_FIRMWARE so they are
built into the kernel image.

Signed-off-by: Stefan Popa <stefan.popa@analog.com>

stefpopa commented:
changelog:
- fixed checkpatch issues
- suppress build warnings squashed into Add ADRV903X support
- firmware: Add ADRV9030 firmware — defconfig split out

stefpopa commented:
changelog:
- fixed clang_analyzer: incompatible integer to pointer conversion initializing 'const char *' with an expression of type 'int'

nunojsa reviewed (approved):
LGTM! I think the CI error is unrelated

gastmaier reviewed (approved):
The ci error was unrelated, failed at checkout.
the gh uptime 98.89 % + external network issues
there is always a chance you get lucky ;) , worth to re-run when this happens
